pkg/admission/admission.go

```diff
@@ -0,0 +1,254 @@
+
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		level.Warn(a.logger).Log("msg", "Unable to deserialize request", "err", err)
+		responseAdmissionReview.Response = toAdmissionResponseFailure("Unable to deserialize request", []error{err})
+	} else {
+		responseAdmissionReview.Response = admit(requestedAdmissionReview)
+	}
+
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+
+	level.Debug(a.logger).Log("msg", "sending response", "content", string(respBytes))
+
+	if err != nil {
+		level.Error(a.logger).Log("msg", "Cannot serialize response", "err", err)
+		http.Error(w, fmt.Sprintf("could not serialize response: %v", err), http.StatusInternalServerError)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		level.Error(a.logger).Log("msg", "Cannot write response", "err", err)
+		http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+	}
+}
+
+func (a *Admission) mutatePrometheusRules(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	level.Debug(a.logger).Log("msg", "Mutating prometheusrules")
+
+	if ar.Request.Resource != ruleResource {
+		err := fmt.Errorf("expected resource to be %v, but received %v", ruleResource, ar.Request.Resource)
+		level.Warn(a.logger).Log("err", err)
+		return toAdmissionResponseFailure("Unexpected resource kind", []error{err})
+	}
+
+	rule := &PrometheusRules{}
+	if err := json.Unmarshal(ar.Request.Object.Raw, rule); err != nil {
+		level.Info(a.logger).Log("msg", "Cannot unmarshal rules from spec", "err", err)
+		return toAdmissionResponseFailure("Cannot unmarshal rules from spec", []error{err})
+	}
+
+	reviewResponse := &v1beta1.AdmissionResponse{Allowed: true}
+	if len(rule.Annotations) == 0 {
+		reviewResponse.Patch = []byte(addFirstAnnotationPatch)
+	} else {
+		reviewResponse.Patch = []byte(addAdditionalAnnotationPatch)
+	}
+	pt := v1beta1.PatchTypeJSONPatch
+	reviewResponse.PatchType = &pt
+	return reviewResponse
+}
+
+func (a *Admission) validatePrometheusRules(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	(*a.validationTriggeredCounter).Inc()
+	level.Debug(a.logger).Log("msg", "Validating prometheusrules")
+
+	if ar.Request.Resource != ruleResource {
+		err := fmt.Errorf("expected resource to be %v, but received %v", ruleResource, ar.Request.Resource)
+		level.Warn(a.logger).Log("err", err)
+		(*a.validationErrorsCounter).Inc()
+		return toAdmissionResponseFailure("Unexpected resource kind", []error{err})
+	}
+
+	rules := &PrometheusRules{}
+	if err := json.Unmarshal(ar.Request.Object.Raw, rules); err != nil {
+		level.Info(a.logger).Log("msg", "Cannot unmarshal rules from spec", "err", err)
+		(*a.validationErrorsCounter).Inc()
+		return toAdmissionResponseFailure("Cannot unmarshal rules from spec", []error{err})
+	}
+
+	_, errors := rulefmt.Parse(rules.Spec.Raw)
```

Comment:
I was under the impression that rules are added here in a completely 1:1 format with Prometheus. If not, then it would become pretty awkward to perform any kind of validation on these before pushing them to the cluster.

Reply:
Yeah for now this is definitely fine.